File: cpp/gabaritos/gabarito25.cpp
#include <iostream>
#include <vector>

void merge(std::vector<int>& lst, int left, int mid, int right) {
    std::vector<int> temp(right - left + 1);
    int i = left, j = mid + 1, k = 0;
    while (i <= mid && j <= right) {
        if (lst[i] <= lst[j]) {
            temp[k++] = lst[i++];
        } else {
            temp[k++] = lst[j++];
        }
    }
    while (i <= mid) temp[k++] = lst[i++];
    while (j <= right) temp[k++] = lst[j++];
    for (int p = 0; p < k; ++p) lst[left + p] = temp[p];
}

void merge_sort(std::vector<int>& lst, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        merge_sort(lst, left, mid);
        merge_sort(lst, mid + 1, right);
        merge(lst, left, mid, right);
    }
}

int main() {
    std::vector<int> lista = {38, 27, 43, 3, 9, 82, 10};
    merge_sort(lista, 0, lista.size() - 1);
    std::cout << "Ordenada: ";
    for (int num : lista) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
    return 0;
}

File: csharp/gabaritos/gabarito25.cs
using System;
using System.Collections.Generic;

class Program
{
    static List<int> MergeSort(List<int> lst)
    {
        if (lst.Count <= 1)
            return lst;
        int mid = lst.Count / 2;
        List<int> left = MergeSort(lst.GetRange(0, mid));
        List<int> right = MergeSort(lst.GetRange(mid, lst.Count - mid));
        return Merge(left, right);
    }

    static List<int> Merge(List<int> left, List<int> right)
    {
        List<int> result = new List<int>();
        int i = 0, j = 0;
        while (i < left.Count && j < right.Count)
        {
            if (left[i] < right[j])
            {
                result.Add(left[i]);
                i++;
            }
            else
            {
                result.Add(right[j]);
                j++;
            }
        }
        result.AddRange(left.GetRange(i, left.Count - i));
        result.AddRange(right.GetRange(j, right.Count - j));
        return result;
    }

    static void Main()
    {
        List<int> lst = new List<int> { 64, 34, 25, 12, 22, 11, 90 };
        List<int> sorted = MergeSort(lst);
        Console.WriteLine("Ordenada: " + string.Join(", ", sorted));
    }
}